apps/rush-lib/src/logic/taskRunner/TaskRunner.ts

```diff
@@ -194,47 +211,75 @@ export class TaskRunner {
    * Removes any non-ready tasks from the build queue (this should only be blocked tasks)
    */
-  private _getNextTask(): Task | undefined {
-    for (let i: number = 0; i < this._buildQueue.length; i++) {
-      const task: Task = this._buildQueue[i];
+  private _getNextTaskSync(): IteratorResult<Task | undefined> {
+    const { _buildQueue: buildQueue } = this;
+
+    for (let i: number = buildQueue.length - 1; i >= 0; i--) {
+      const task: Task = buildQueue[i];
 
       if (task.status !== TaskStatus.Ready) {
         // It shouldn't be on the queue, remove it
-        this._buildQueue.splice(i, 1);
-        // Decrement since we modified the array
-        i--;
+        // This should be a blocked task
+        buildQueue.splice(i, 1);
       } else if (task.dependencies.size === 0 && task.status === TaskStatus.Ready) {
         // this is a task which is ready to go. remove it and return it
-        return this._buildQueue.splice(i, 1)[0];
+        return {
+          done: false,
+          value: buildQueue.splice(i, 1)[0]
+        };
       }
       // Otherwise task is still waiting
     }
-    return undefined; // There are no tasks ready to go at this time
+
+    return {
+      done: buildQueue.length === 0,
+      value: undefined
+    }; // There are no tasks ready to go at this time
   }
 
-  /**
-   * Helper function which finds any tasks which are available to run and begins executing them.
-   * It calls the complete callback when all tasks are completed, or rejects if any task fails.
-   */
-  private async _startAvailableTasksAsync(): Promise<void> {
-    const taskPromises: Promise<void>[] = [];
-    let ctask: Task | undefined;
-    while (this._currentActiveTasks < this._parallelism && (ctask = this._getNextTask())) {
-      this._currentActiveTasks++;
-      const task: Task = ctask;
-      task.status = TaskStatus.Executing;
-
-      task.stopwatch = Stopwatch.start();
-      task.collatedWriter = this._streamCollator.registerTask(task.name);
-      task.stdioSummarizer = new StdioSummarizer();
-
-      taskPromises.push(this._executeTaskAndChainAsync(task));
-    }
+  private _getTaskIterable(): AsyncIterable<Task> {
+    return {
+      [Symbol.asyncIterator]: (): AsyncIterator<Task, void, undefined> => {
+        return {
+          next: (): Promise<IteratorResult<Task, void>> => {
+            const syncResult: IteratorResult<Task | undefined> = this._getNextTaskSync();
+
+            if (syncResult.value || syncResult.done) {
+              return Promise.resolve(syncResult as IteratorResult<Task, void>);
+            }
+
+            return new Promise(
+              (resolve: (result: IteratorResult<Task, void>) => void, reject: (err: Error) => void) => {
+                this._waitingWorkers.push([resolve, reject]);
+              }
+            );
+          }
+        };
+      }
+    };
+  }
+
+  private async _taskWorker(tid: number): Promise<void> {
```

Comment:
```suggestion
  private async _taskWorkerAsync(tid: number): Promise<void> {
```

Reply:
Refactored this into a separate class that handles async task queue iteration.